### PR TITLE
Port some Moonbeam precompiles

### PR DESCRIPTION
**Pull request by fgamundi**
Ports some EVM precompiled smart contracts from Moonbeam:
- pallet-balances
- batch
- call-permit
- xcm-utils

### CODE REVIEW

**Review by girazoki — APPROVED** [review]
LGTM, I think we should be able to merge this now right?

**Comment by librelois** [comment]
Wait, I need to review it before

File: precompiles/balances-erc20/ERC20.sol [hunk 1]
```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity >=0.8.3;
+
+/// @dev The IERC20 contract's address.
+address constant IERC20_ADDRESS = 0x0000000000000000000000000000000000000802;
```
Comment (librelois) [inline precompiles/balances-erc20/ERC20.sol:5]: Theses addresses are specific to moonbeam, so if we move these solidity interfaces to moonkit we should use a template/placeholder for theses addresses

Comment (fgamundi) [reply precompiles/balances-erc20/ERC20.sol:5]: You're right. I think we can also remove them, as I don't think they're being used

Comment (librelois) [reply precompiles/balances-erc20/ERC20.sol:5]: I think there are useful, I'm used them several time to figure out what precompile a transaction call from the address.
You cna remove them from the solidity intreface but I would be great to have a tool that generate a table of precompiles addresses, @nanocryk IIRC we discuss in the past to create such a tool, what was the blockers?

Comment (fgamundi) [reply precompiles/balances-erc20/ERC20.sol:5]: I agree that they're useful, but doesn't make sense to keep the address in the solidity contract in moonkit, as the address will change depending on the runtime using it. Maybe we can have the list of addresses somewhere else, like a comment in the runtime's `precompiles.rs`?

Comment (librelois) [reply precompiles/balances-erc20/ERC20.sol:5]: > Maybe we can have the list of addresses somewhere else, like a comment in the runtime's precompiles.rs?

Yes, that's a good idea

**Review by Agusrodri — APPROVED** [review]
Looks good to me

**Pull request merged**